### PR TITLE
Fix SkipUntil to not call MoveNext past end of sequence

### PR DESCRIPTION
**Pull request by Orace**
~Fix a part of #649~
Fork of #660 

SkipUntil called MoveNext after sequence end

### CODE REVIEW

**Comment by Orace** [comment]
> How is this related to #649?

It is not, sorry I copy paste to quickly. I updated the first comment..

It's a re-call MoveNext on sequence end

**Comment by atifaziz** [comment]
So this is the fix for issue #664?

**Comment by Orace** [comment]
> So this is the fix for issue #664?

Indeed, I forgot to mention it :confused:

File: MoreLinq.Test/SkipUntilTest.cs [hunk 1]
```diff
@@ -57,5 +59,25 @@ public void SkipUntilEvaluatesPredicateLazily()
             var sequence = Enumerable.Range(-2, 5).SkipUntil(x => 1 / x == -1);
             sequence.AssertSequenceEqual(0, 1, 2);
         }
+
+        public static readonly IEnumerable<ITestCaseData> TestData =
+            from e in new[]
+            {
+                new {s = new int[0]     , o = 0, r = new int[0]},   // empty sequence
```
Comment (atifaziz) [inline MoreLinq.Test/SkipUntilTest.cs]: Can we use more meaningful (Pascal-case) names for the properties?

```diff
-                new {s = new int[0]     , o = 0, r = new int[0]},   // empty sequence
+                new { Source = new int[0]     , Min = 0, Expected = new int[0] },   // empty sequence
```

Also, space after `{` and before `}`.

File: MoreLinq/SkipUntil.cs [hunk 2]
```diff
@@ -61,14 +61,19 @@ public static IEnumerable<TSource> SkipUntil<TSource>(this IEnumerable<TSource>
             {
                 using var iterator = source.GetEnumerator();
 
-                while (iterator.MoveNext())
+                for(;;)
```
Comment (atifaziz) [inline MoreLinq/SkipUntil.cs]:
```suggestion
                for (;;)
```

File: MoreLinq/SkipUntil.cs [hunk 3]
```diff
@@ -61,14 +61,19 @@ public static IEnumerable<TSource> SkipUntil<TSource>(this IEnumerable<TSource>
...
                 {
+                    if (!iterator.MoveNext())
+                        yield break;
+
                     if (predicate(iterator.Current))
                         break;
                 }
 
                 while (iterator.MoveNext())
+                {
```
Comment (atifaziz) [inline MoreLinq/SkipUntil.cs]: Seems inconsistent with above so remove:

```suggestion
```

File: MoreLinq/SkipUntil.cs [hunk 4]
```diff
@@ -61,14 +61,19 @@ public static IEnumerable<TSource> SkipUntil<TSource>(this IEnumerable<TSource>
...
                     yield return iterator.Current;
+                }
```
Comment (atifaziz) [inline MoreLinq/SkipUntil.cs]: Seems inconsistent with above so remove:

```suggestion
```

File: MoreLinq/SkipUntil.cs (hunk 2 above)
Comment (Orace) [reply MoreLinq/SkipUntil.cs]: It also can be a. do/while loop.
Anyway, never to late to ask, is there any formating rules description somewhere ? I use resharper for auto-formating and I didn't customize the settings...
In the same fashion of questions : `for(;;)` or `while(true)`

File: MoreLinq/SkipUntil.cs [hunk 5]
```diff
@@ -59,16 +59,16 @@ public static IEnumerable<TSource> SkipUntil<TSource>(this IEnumerable<TSource>
...
 
-                while (iterator.MoveNext())
+                do
                 {
-                    if (predicate(iterator.Current))
-                        break;
-                }
+                    if (!enumerator.MoveNext())
+                        yield break;
+                } while(!predicate(enumerator.Current))
```
Comment (Orace) [inline MoreLinq/SkipUntil.cs]:
```suggestion
                } while(!predicate(enumerator.Current));
```

**Comment by Orace** [comment]
A squash merge may reduce my shame

File: MoreLinq/SkipUntil.cs (hunk 2 above)
Comment (atifaziz) [reply MoreLinq/SkipUntil.cs]: Nothing document but hope fd32a8ad5754aea1288c277dedb927053e1c8368 helps as a starting point. Other than that, readability comes first (even if it means breaking rules), then consistency and finally general C# guidelines. Unfortunately, this means I have to switch off all auto-formatting in the IDE; prefer to write for humans, not machines.

> In the same fashion of questions : `for(;;)` or `while(true)` ?

Either is fine as long as there is as space after `for` or `while`:

https://github.com/morelinq/MoreLINQ/blob/fd32a8ad5754aea1288c277dedb927053e1c8368/.editorconfig#L51

**Pull request merged**